pydicom/filereader.py

```diff
@@ -282,6 +294,12 @@ def data_element_generator(fp,
         if length != 0xFFFFFFFF:
             # don't defer loading of Specific Character Set value as it is
             # needed immediately to get the character encoding for other tags
+            if tag_set is not None and tag not in tag_set and tag != (
```

Comment:
Another thought:  could pre-calc the comparison:  `have_tag_set = tag_set is not None".  Easier to read and take about 60% of time in timeit.